File: cpp/src/ch05/time_keeper.cc

#include "time_keeper.h"

bool TimeKeeper::is_time_over() const
{
    using std::chrono::duration_cast;
    using std::chrono::milliseconds;

    auto diff = std::chrono::high_resolution_clock::now() - this->start_time_;
    return duration_cast<milliseconds>(diff).count() >= time_threshold_;
}

float TimeKeeper::get_elapsed_time()
{
    using std::chrono::duration_cast;
    using std::chrono::milliseconds;

    auto diff = std::chrono::high_resolution_clock::now() - this->start_time_;
    return duration_cast<milliseconds>(diff).count();
}

File: python/src/thunder_book/ch05/time_keeper.py
from datetime import datetime


class TimeKeeper:
    def __init__(self, time_threshold: int) -> None:
        """_summary_

        Args:
            time_threshold (int): threshold time in milliseconds
        """
        self.time_threshold = time_threshold
        self.start_time = datetime.now()

    def time_diff(self):
        diff = datetime.now() - self.start_time
        return diff

    def is_time_over(self) -> bool:
        diff = datetime.now() - self.start_time
        return diff.total_seconds() * 1000 > self.time_threshold
